src/analyses/goto_check.cpp

```diff
@@ -1897,39 +1938,96 @@ void goto_checkt::goto_check(
+    std::list<bool *> already_checked;
     const auto &pragmas = i.source_location().get_pragmas();
     for(const auto &d : pragmas)
     {
       if(d.first == "disable:bounds-check")
-        flag_resetter.set_flag(enable_bounds_check, false);
+        resetter.set_flag(enable_bounds_check, false, d.first);
       else if(d.first == "disable:pointer-check")
-        flag_resetter.set_flag(enable_pointer_check, false);
+        resetter.set_flag(enable_pointer_check, false, d.first);
       else if(d.first == "disable:memory-leak-check")
-        flag_resetter.set_flag(enable_memory_leak_check, false);
+        resetter.set_flag(enable_memory_leak_check, false, d.first);
       else if(d.first == "disable:div-by-zero-check")
-        flag_resetter.set_flag(enable_div_by_zero_check, false);
+        resetter.set_flag(enable_div_by_zero_check, false, d.first);
       else if(d.first == "disable:enum-range-check")
-        flag_resetter.set_flag(enable_enum_range_check, false);
+        resetter.set_flag(enable_enum_range_check, false, d.first);
       else if(d.first == "disable:signed-overflow-check")
-        flag_resetter.set_flag(enable_signed_overflow_check, false);
+        resetter.set_flag(enable_signed_overflow_check, false, d.first);
       else if(d.first == "disable:unsigned-overflow-check")
-        flag_resetter.set_flag(enable_unsigned_overflow_check, false);
+        resetter.set_flag(enable_unsigned_overflow_check, false, d.first);
       else if(d.first == "disable:pointer-overflow-check")
-        flag_resetter.set_flag(enable_pointer_overflow_check, false);
+        resetter.set_flag(enable_pointer_overflow_check, false, d.first);
       else if(d.first == "disable:float-overflow-check")
-        flag_resetter.set_flag(enable_float_overflow_check, false);
+        resetter.set_flag(enable_float_overflow_check, false, d.first);
       else if(d.first == "disable:conversion-check")
-        flag_resetter.set_flag(enable_conversion_check, false);
+        resetter.set_flag(enable_conversion_check, false, d.first);
       else if(d.first == "disable:undefined-shift-check")
-        flag_resetter.set_flag(enable_undefined_shift_check, false);
+        resetter.set_flag(enable_undefined_shift_check, false, d.first);
       else if(d.first == "disable:nan-check")
-        flag_resetter.set_flag(enable_nan_check, false);
+        resetter.set_flag(enable_nan_check, false, d.first);
       else if(d.first == "disable:pointer-primitive-check")
-        flag_resetter.set_flag(enable_pointer_primitive_check, false);
+        resetter.set_flag(enable_pointer_primitive_check, false, d.first);
+      else if(d.first == "enable:bounds-check")
+        resetter.set_flag(enable_bounds_check, true, d.first);
+      else if(d.first == "enable:pointer-check")
+        resetter.set_flag(enable_pointer_check, true, d.first);
+      else if(d.first == "enable:memory_leak-check")
+        resetter.set_flag(enable_memory_leak_check, true, d.first);
```

Comment:
Randomly placing my comment in this block of code: I think we should now look at refactoring this piece of code, perhaps as follows:
1) string-split `id2string(d.first)` on `:`
2) parse the first component into an enum { DISABLE, ENABLED, CHECKED }
3) lookup the second component in a map that is initialised in the constructor (the map will map strings or `irep_idt`to `bool*` or `bool&`)
4) perform a `set_flag` operation on the result of that lookup, directed by the enum value.

As is, this will become a maintenance nightmare.

Reply:
Yes, agreed.